// nexe/lib/steps/shim-fs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var assert_1 = require("assert");
var path_1 = require("path");
var binary = process.__nexe;
assert_1.ok(binary);
var manifest = binary.resources;
var directories = {};
var isString = function (x) { return typeof x === 'string' || x instanceof String; };
var isNotFile = function () { return false; };
var isNotDirectory = isNotFile;
var isFile = function () { return true; };
var isDirectory = isFile;
if (Object.keys(manifest).length) {
    var fs_1 = require('fs');
    var originalReadFile_1 = fs_1.readFile;
    var originalReadFileSync_1 = fs_1.readFileSync;
    var originalReaddir_1 = fs_1.readdir;
    var originalReaddirSync_1 = fs_1.readdirSync;
    var originalStatSync_1 = fs_1.statSync;
    var originalStat_1 = fs_1.stat;
    var resourceStart_1 = binary.layout.resourceStart;
    var statTime_1 = function () {
        var stat = binary.layout.stat;
        return {
            dev: 0,
            ino: 0,
            nlink: 0,
            rdev: 0,
            uid: 123,
            gid: 500,
            blksize: 4096,
            blocks: 0,
            atime: new Date(stat.atime),
            atimeMs: stat.atime.getTime(),
            mtime: new Date(stat.mtime),
            mtimeMs: stat.mtime.getTime(),
            ctime: new Date(stat.ctime),
            ctimMs: stat.ctime.getTime(),
            birthtime: new Date(stat.birthtime),
            birthtimeMs: stat.birthtime.getTime()
        };
    };
    var createStat_1 = function (directoryExtensions, fileExtensions) {
        if (!fileExtensions) {
            return Object.assign({}, binary.layout.stat, directoryExtensions, { size: 0 }, statTime_1());
        }
        var size = directoryExtensions[1];
        return Object.assign({}, binary.layout.stat, fileExtensions, { size: size }, statTime_1());
    };
    var ownStat_1 = function (path) {
        var key = path_1.resolve(path);
        if (directories[key]) {
            return createStat_1({ isDirectory: isDirectory, isFile: isNotFile });
        }
        if (manifest[key]) {
            return createStat_1(manifest[key], { isFile: isFile, isDirectory: isNotDirectory });
        }
    };
    var setupManifest_1 = function () {
        Object.keys(manifest).forEach(function (key) {
            var absolutePath = path_1.resolve(key);
            var dirPath = path_1.dirname(absolutePath);
            directories[dirPath] = directories[dirPath] || {};
            directories[dirPath][path_1.basename(absolutePath)] = true;
            if (!manifest[absolutePath]) {
                manifest[absolutePath] = manifest[key];
            }
            var normalizedPath = path_1.normalize(key);
            if (!manifest[normalizedPath]) {
                manifest[normalizedPath] = manifest[key];
            }
        });
        setupManifest_1 = function () { };
    };
    //naive patches intended to work for most use cases
    var nfs = {
        readdir: function readdir(path, options, callback) {
            setupManifest_1();
            path = path.toString();
            if ('function' === typeof options) {
                callback = options;
                options = { encoding: 'utf8' };
            }
            var dir = directories[path_1.resolve(path)];
            if (dir) {
                process.nextTick(function () {
                    callback(null, Object.keys(dir));
                });
            }
            else {
                return originalReaddir_1.apply(fs_1, arguments);
            }
        },
        readdirSync: function readdirSync(path, options) {
            setupManifest_1();
            path = path.toString();
            var dir = directories[path_1.resolve(path)];
            if (dir) {
                return Object.keys(dir);
            }
            return originalReaddirSync_1.apply(fs_1, arguments);
        },
        readFile: function readFile(file, options, callback) {
            setupManifest_1();
            var entry = manifest[file];
            if (!entry || !isString(file)) {
                return originalReadFile_1.apply(fs_1, arguments);
            }
            var offset = entry[0], length = entry[1];
            var resourceOffset = resourceStart_1 + offset;
            var encoding = isString(options) ? options : null;
            callback = typeof options === 'function' ? options : callback;
            fs_1.open(process.execPath, 'r', function (err, fd) {
                if (err)
                    return callback(err, null);
                fs_1.read(fd, Buffer.alloc(length), 0, length, resourceOffset, function (error, bytesRead, result) {
                    if (error) {
                        return fs_1.close(fd, function () {
                            callback(error, null);
                        });
                    }
                    fs_1.close(fd, function (err) {
                        if (err) {
                            return callback(err, result);
                        }
                        callback(err, encoding ? result.toString(encoding) : result);
                    });
                });
            });
        },
        readFileSync: function readFileSync(file, options) {
            setupManifest_1();
            var entry = manifest[file];
            if (!entry || !isString(file)) {
                return originalReadFileSync_1.apply(fs_1, arguments);
            }
            var offset = entry[0], length = entry[1];
            var resourceOffset = resourceStart_1 + offset;
            var encoding = isString(options) ? options : null;
            var fd = fs_1.openSync(process.execPath, 'r');
            var result = Buffer.alloc(length);
            fs_1.readSync(fd, result, 0, length, resourceOffset);
            fs_1.closeSync(fd);
            return encoding ? result.toString(encoding) : result;
        },
        statSync: function statSync(path) {
            var stat = ownStat_1(path);
            if (stat) {
                return stat;
            }
            return originalStatSync_1.apply(fs_1, arguments);
        },
        stat: function stat(path, callback) {
            var stat = ownStat_1(path);
            if (stat) {
                process.nextTick(function () {
                    callback(null, stat);
                });
            }
            else {
                return originalStat_1.apply(fs_1, arguments);
            }
        }
    };
    Object.assign(fs_1, nfs);
}
